qermit/clifford_noise_characterisation/ccl.py

```diff
@@ -295,12 +314,36 @@ def task(
             # generate all state circuits
             c_copy = ansatz_circuit.Circuit.copy()
             c_copy.symbol_substitution(ansatz_circuit.SymbolsDict._symbolic_map)
-            state_circuits = gen_state_circuits(
-                c_copy,
-                n_non_cliffords,
-                n_pairs,
-                total_state_circuits,
-            )
+
+            all_close = True
+            attempt = 0
+            while all_close and attempt < max_state_circuits_attempts:
+
+                state_circuits = gen_state_circuits(
```

Comment:
Would it make more sense to check each state circuit individually and then rebuild if it isn't close?
I guess the advantage here is that it would reassign the "Non-Clifford" and "Clifford" rz ops, meaning it could eventually find a distribution which is not close?

Reply:
Actually likely that wouldn't do that - should we add a check for each circuit? Or would you rather we do that later and just get on with it for now?

Reply:
The check here isolates if all expectation values are close, and rejects the list if they're all close. As such we would have to generate all of the circuits to check if at least one is different from the rest (the one that is different could be the last one we generate, for example). An alternative to regenerating the whole list which I can think of might be to generate the whole list once and then regenerate and replace individual circuits in the list until they are different from the rest. The reasons I chose not to do that are:
- It's a little more honest to the original proposal to fix "Non-Clifford" and "Clifford" rz ops and generate lists accordingly, as far as I understand.
- Restarting the whole process will select new sets of "Non-Clifford" and "Clifford" rz ops, which avoids the problem of getting stuck with sets of "Non-Clifford" and "Clifford" rz ops which generate circuits with similar expectation values.
- Regenerating the whole list might give a list with a few circuits with different expectation values, which is even better than one being different.
You'll certainly know more about whether the first two points are of significant concern, so you can correct me if I missed the point.

I can say that the approach implemented does seem to work well with the benchmarking circuits, and regenerating the whole list is not a significant time sink. Since it works well for our practical purposes I'm okay with coming back later with new ideas but I'm happy to debate this :P .